### ISSUE TITLE
Pre-RFC: Multi-step transactions

### ISSUE THREAD

**Issue by dunnock** [tari-project/validator-node#10]
By example of sell token, when transfer of ownership should happen after a given event, e.g. tari's in a specific Wallet. Here I will just paste message from @Timo614 from chat:

During the meeting Max brought up the situation where the desire exists to transfer a token. We discussed potentially having the initiating node (the one contacted) handle the transaction prior to broadcasting the instruction or sending it off to the leader to handle that as part of their own processing prior to the leader broadcasting it out (although the leader will likely change between consensus periods so I think the former is ideal). 

Either choice one issue though is that when you begin the transfer process you really don't want other instructions going through that may alter the state of the token. In addition you could sell your token to multiple people and while it's waiting on those wallet payments to clear the logic will not know to prevent it. To take it further you could have multiple nodes that each received a transfer instruction, each begins the process of creating a dummy wallet to take payment, etc. and then they all broadcast out they received payment so the instruction to transfer the token goes out which will fail.

Wondering now... maybe we need some multiple step process where a node uses an instruction to place a timed lock on a token where only instructions sourced by that node can be run. If all nodes agree they place that lock on the token. When the initiating node sees that the lock exists, it goes and handles the dummy wallet, etc. and once it transfers it the lock is removed. All nodes would validate during instruction processing whether the source node for the instruction was the one that gave the lock? The lock would auto expire after some period to prevent tokens from staying in a locked state if a node dropped off after it asked for the lock.

So:

- Initiating node receives instruction to transfer token
- Initiating node submits instruction for a timed lock to the committee
- Upon the timed lock instruction being approved and included in the current state (via the consensus mechanism) the initiating node now can perform transfer instructions so it creates a dummy wallet that awaits transfer for the fee
- Upon confirmations in the base layer of the transfer the initiating node takes the initial instruction and it propagates it
- Other node members confirm that the request came from the initiating node that currently has a lock as part of their own processing of instructions for that token if a lock exists (as it does in this case)
- Instruction handled normally by the consensus process but upon completion of the transfer it removes the lock (or a second instruction to remove the timed lock can be sent and approved as part of the consensus process)
- At some point the timed lock is automatically removed because it passes its expiration window and normal instructions can be run on the token again if the transfer does not occur or no request is made to remove the lock

**Comment by dunnock** [comment]
@Timo614 on your proposed solution above that all seems cool, though I am not sure that node need to wait for timed lock to confirm before proposing wallet ID, creation of temp wallet is cheap and will make user able to proceed with payment not waiting for consensus on lock. Let's call your proposed solution __Distributed transaction lock__ or whatever other name you suggest? I will post few more options below

**Comment by dunnock** [comment]
__Option 2 - multiple transactions and distinct token states__:
- Initiating node receives instruction, submits first transaction to move token to `Locked` ( `Active` ) state with timeout, at the same time providing `temp wallet` key for payment
- Considering payment received within timeout:
- As soon as payment received initiating node starting 2nd transaction moving token to `Active` state with `new owner`, also distributing tari's to [`beneficiary wallets`] (including `old owner` and whoever else is receiving fee on asset transactions)
- Consensus on template includes validation that both incoming transaction to `temp wallet` as well as outgoing transfers to [`beneficiary wallets`] happened as well as ownership and other token validation succeeded
- If any of steps above fail initiator node has to cancel or revert all transfers (depending if tari has a way to cancel transfer - I was it in wallet app) otherwise it will be considered malicious node and might lose collateral
- If initiating node accepts tari on temp wallet and does not move - it will be considered malicious node and might lose collateral

Pros:
- 2 short transactions
- initiating node secures transfers by it's collateral

Cons:
- initiator node bound to token via wallet
- initiator node has to stay online until 2nd transaction completes, otherwise might lose collateral

**Comment by dunnock** [comment]
__Option 3 - one final state transaction__
I believe we touched this option on a call, it might work as well I guess:
- initiator node receives instruction, provides `temp wallet` and holding it's state until receives amount on `temp wallet`
- if no payment received initiator node will revert payment
- as soon as payment received initiating node starting 2nd transaction moving token to Active state with new owner, also distributing tari's to [beneficiary wallets] (including old owner and whoever else is receiving fee on asset transactions)
- consensus on transaction happens same as with __Option 2__
- If operation on the same token has happened in parallel (e.g. double spend) we can cancel both transactions (or accept one - perhaps leader's decision)

Pros:
- simplified flow, short transaction

Cons:
- it feel like initiator node might cheat without losing collateral...

**Comment by dunnock** [comment]
__Option 1 - distributed transaction lock__ pros and cons from my POV:

Pros:
- tari's transfer is secured by VN's collateral as wallet was communicated to committee - it feels more secure than Option 2

Cons:
- user will have to wait for consensus until being able to make payment, if that payment requires user's consent that might be online waiting time
- longer transaction with lock

**Comment by dunnock** [comment]
__Option 4 - distinct token states with payment processor__: -- similar to option 2
To secure payment we might split initiator node and payment processor node. Initiator node before posting transaction will have to acquire temp wallet of payment processor node (e.g. consider current leader).
- Initiating node receives instruction, submits first transaction to move token to `Locked` ( `Active` ) state with timeout, at the same time providing `temp wallet` key for payment of payment processor node
- Considering payment received within timeout:
- As soon as payment received payment processor node starting 2nd transaction moving token to `Active` state with `new owner`, also distributing tari's to [`beneficiary wallets`] (including `old owner` and whoever else is receiving fee on asset transactions)
- Consensus on template includes validation that both incoming transaction to `temp wallet` as well as outgoing transfers to [`beneficiary wallets`] happened as well as ownership and other token validation succeeded
- If any of steps above fail payment processor node has to cancel or revert all transfers (depending if tari has a way to cancel transfer - I was it in wallet app) otherwise it will be considered malicious node and might lose collateral

Pros:
- 2 short transactions
- payment processor node secures transfers by it's collateral
- it is harder for initiator node to cheat with wallet when payment processor is another random node

Cons:
- initiator node bound to token via wallet
- initiator node has to stay online until 2nd transaction completes, otherwise might lose collateral

**Comment by Timo614** [comment]
> @Timo614 on your proposed solution above that all seems cool, though I am not sure that node need to wait for timed lock to confirm before proposing wallet ID, creation of temp wallet is cheap and will make user able to proceed with payment not waiting for consensus on lock. Let's call your proposed solution Distributed transaction lock or whatever other name you suggest? I will post few more options below

Yeah the thinking there was that by waiting for the timed lock before proposing wallet ID we prevent the situation where multiple nodes receive a transfer request in tandem for the same token, all allow payments of fees, and then only the first to be approved for the lock is able to complete the transfer. I guess switching that would work as long as it reverts the transaction when the subsequent node instructions see that their lock failed.

Not sure how realistic that would be in practice though.

**Comment by dunnock** [comment]
> Yeah the thinking there was that by waiting for the timed lock before proposing wallet ID we prevent the situation where multiple nodes receive a transfer request in tandem for the same token, all allow payments of fees, and then only the first to be approved for the lock is able to complete the transfer. I guess switching that would work as long as it reverts the transaction when the subsequent node instructions see that their lock failed.

Yeah, so this __Option 1__ is more secure than __Option 2__ it seems, though from the point that it makes impossible for the VN to cheat (I think we will need to include wallet in the Lock so that other nodes can spot if any transaction was made to that wallet indeed). __Option 3__ is the least secure. I just thought of __Option 4__ above where initiator node does not provide it's wallet rather acquires wallet from leader (or other temp node) - it seems not least secure than __Option 1__.

I think if malicious token holder will try to sell same token to different people we are ok to cancel all transactions and revert all transfers.

**Comment by dunnock** [comment]
Added diagrams to illustrate __Option 1__ and __Option 4__
